src/display/node_stream.js

```diff
@@ -0,0 +1,384 @@
+    this._readCapability = createPromiseCapability();
+
+    this._isStreamingSupported = !stream.source.disableStream;
+  }
+
+  get isStreamingSupported() {
+    return this._isStreamingSupported;
+  }
+}
+
+class PDFNodeStreamFullReader extends BaseFullReader {
+  constructor(stream) {
+    super(stream);
+
+    this._disableRange = stream.options.disableRange || false;
+    this._rangeChunkSize = stream.source.rangeChunkSize;
+    if (!this._rangeChunkSize && !this._disableRange) {
+      this._disableRange = true;
+    }
+
+    let options = {
+      host: this._url.host,
+      path: this._url.path,
+      method: 'GET',
+      headers: stream.httpHeaders,
+    };
+
+    let handleResponse = (response) => {
+      this._headersCapability.resolve();
+      this._readableStream = response;
+
+      response.on('readable', () => {
+        this._readCapability.resolve();
+      });
+
+      response.on('end', () => {
+        // Destroy response to minimize resource usage.
+        response.destroy();
+        this._done = true;
+        this._readCapability.resolve();
+      });
+
+      response.on('error', (reason) => {
+        this._errored = true;
+        this._reason = reason;
+        this._readCapability.resolve();
+      });
+
+      let { allowRangeRequests, suggestedLength, } =
+      validateRangeRequestCapabilities({
+        getResponseHeader: this.getResponseHeader.bind(this),
+        isHttp: stream.isHttp,
+        rangeChunkSize: this._rangeChunkSize,
+        disableRange: this._disableRange,
+      });
+
+      if (allowRangeRequests) {
+        this._isRangeSupported = true;
+      }
+      this._length = suggestedLength;
+    };
+
+    this._request = this._url.protocol === 'http:' ?
+      http.request(options, handleResponse) :
+      https.request(options, handleResponse);
+
+    this._request.on('error', (reason) => {
+      this._errored = true;
+      this._reason = reason;
+      this._headersCapability.reject(reason);
+    });
+    this._request.end();
+  }
+
+  getResponseHeader(name) {
+    return this._readableStream.headers[name];
+  }
+}
+
+class PDFNodeStreamRangeReader extends BaseRangeReader {
+  constructor(stream, start, end) {
+    super(stream);
+
+    this._readableStream = null;
+    let rangeStr = start + '-' + (end - 1);
+    stream.httpHeaders['Range'] = 'bytes=' + rangeStr;
+
+    let options = {
+      host: this._url.host,
+      path: this._url.path,
+      method: 'GET',
+      headers: stream.httpHeaders,
+    };
+    let handleResponse = (response) => {
+      this._readableStream = response;
+
+      response.on('readable', () => {
+        this._readCapability.resolve();
+      });
+
+      response.on('end', () => {
+        response.destroy();
+        this._done = true;
+        this._readCapability.resolve();
+      });
+
+      response.on('error', (reason) => {
+        this._errored = true;
+        this._reason = reason;
+        this._readCapability.resolve();
+      });
```

Comment:
move all `.on('readable'/'end'/'error',` into  e.g. private `_setReadableStream` into base class -- it can be use by Fs constructor as well.

The same thing needs to be done for "full" readers too.